apps/survey/contracts/Survey.sol

```diff
@@ -98,84 +99,104 @@ contract Survey is AragonApp {
+        for (uint256 j = 1; j <= _optionIds.length; j++) {
+            // parameter arrays don't have abstain vote, while survey ones do, so we subtract 1 here to the index
+            uint256 optionId = _optionIds[j-1];
+            uint256 stake = _stakes[j-1];
+            require(optionId != ABSTAIN_VOTE && optionId <= survey.options);
+            require(stake > 0);
 
             // register voter amount
-            // index 0 in options array is reserved for NO_VOTE, so index will be 1 more
-            survey.voters[msg.sender].castedVotes[j + 1] = OptionCast({optionId: _options[j], stake: _stakes[j]});
+            // index 0 in options array is reserved for ABSTAIN_VOTE, so index will be 1 more
+            survey.votes[msg.sender].castedVotes[j] = OptionCast({optionId: optionId, stake: stake});
 
             // add to total option support
-            survey.votes[_options[j]] = survey.votes[_options[j]].add(_stakes[j]);
+            survey.optionVotes[optionId] = survey.optionVotes[optionId].add(stake);
 
             // let's avoid repeating an option by
             // making sure that ascending order is preserved in options array
-            // index 0 in options array is reserved for NO_VOTE, so here we are comparing
-            // the option just added (_options[j] = options[j+1]) with the previous one (options[j])
-            require(survey.voters[msg.sender].castedVotes[j].optionId < survey.voters[msg.sender].castedVotes[j + 1].optionId);
-            totalVoted = totalVoted.add(_stakes[j]);
+            // index 0 in options array is reserved for ABSTAIN_VOTE, so here we are comparing
+            // the option just added with the previous one
+            require(survey.votes[msg.sender].castedVotes[j - 1].optionId < survey.votes[msg.sender].castedVotes[j].optionId);
+            totalVoted = totalVoted.add(stake);
 
-            CastVote(_surveyId, msg.sender, _options[j], _stakes[j]);
+            CastVote(_surveyId, msg.sender, optionId, stake);
         }
 
         // compute and register non used tokens
         // implictly we are doing require(totalVoted <= voterStake) too
-        // (as stated before, index 0 is for NO_VOTE option)
-        survey.voters[msg.sender].castedVotes[0].stake = voterStake.sub(totalVoted);
+        // (as stated before, index 0 is for ABSTAIN_VOTE option)
+        uint256 voterStake = token.balanceOfAt(msg.sender, survey.snapshotBlock);
+        survey.votes[msg.sender].castedVotes[0].stake = voterStake.sub(totalVoted);
 
         // register number of options voted
-        survey.voters[msg.sender].optionsCastedLength = _options.length;
+        survey.votes[msg.sender].optionsCastedLength = _optionIds.length;
 
         // add voter tokens to participation
         survey.participation = survey.participation.add(totalVoted);
+        require(survey.participation <= survey.votingPower);
```

Comment:
What do you think about making this an `assert()`? If this isn't true, I feel like something horrible has gone wrong (rather than input checking).